src/Inference/SIM.hs

```diff
@@ -1,74 +1,60 @@
-{-# LANGUAGE DataKinds #-}
-{-# LANGUAGE FlexibleContexts #-}
-{-# LANGUAGE GADTs #-}
-{-# LANGUAGE PatternSynonyms #-}
-{-# LANGUAGE ScopedTypeVariables #-}
-{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE DataKinds                  #-}
+{-# LANGUAGE FlexibleContexts           #-}
+{-# LANGUAGE GADTs                      #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
 {-# OPTIONS_GHC -Wno-incomplete-patterns #-}
+{-# LANGUAGE FlexibleInstances          #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE KindSignatures             #-}
+{-# LANGUAGE MultiParamTypeClasses      #-}
+{-# LANGUAGE TypeOperators              #-}
+{-# LANGUAGE UndecidableInstances #-}
 
 {- | Simulation.
 -}
 
 module Inference.SIM (
     simulate
-  , runSimulate
-  , traceSamples
-  , handleSamp
-  , handleObs) where
-
-import Data.Map (Map)
-import Effects.Dist ( Observe(..), Sample(..), Dist )
-import Effects.ObsReader ( ObsReader )
-import Effects.State ( State, modify, handleState )
-import Env ( Env )
-import Model ( Model, handleCore )
-import OpenSum (OpenSum)
-import PrimDist ( pattern PrimDistPrf, sample )
-import Prog ( Member(prj), Prog(..), discharge )
-import qualified Data.Map as Map
-import qualified OpenSum
-import Sampler ( Sampler )
-import Trace ( FromSTrace(..), STrace, updateSTrace )
-import Unsafe.Coerce (unsafeCoerce)
+  , runSimulate, DistC, runDist) where
+import           Control.Algebra              (Has, Algebra (alg), (:+:) (L, R))
+import           Control.Carrier.Draw         (DrawC, runDraw)
+import           Control.Carrier.ObsReader    (ObsReaderC, runObsReader)
+import           Control.Carrier.SampleTracer (SampleTracerC, runSampleTracer)
+import           Control.Effect.Dist          (Dist (..))
+import qualified Data.Map                     as Map
+import           Data.Maybe                   (fromMaybe)
+import           Env                          (Env)
+import           PrimDist                     (draw)
+import           Sampler                      (Sampler)
+import           Trace                        (FromSTrace (fromSTrace), STrace)
+import Control.Effect.Lift (Lift)
+import Control.Carrier.Lift (LiftC, runM)
+import Control.Effect.Lift (sendM)
+
+-- Dist Effect Carrier
+newtype DistC (m :: * -> *) (k :: *) = DistC { runDist :: m k }
```

Comment:
This name collides with the default `DistC` carrier, it should be descriptively different.